src/components/heartbeats/heartbeatsStore.ts

```diff
@@ -0,0 +1,148 @@
+
+import { Tx, LeapTransaction } from 'leap-core';
+
+type SlotHeartbeatsData = {
+  periods: Set<number>;
+  lastPeriodChecked: number;
+  owner: string;
+};
+
+export class HeartbeatsStore {
+  private storageKey: string;
+
+  @observable
+  private heartbeatColor: number;
+
+  public keepPeriodsLimit: number;
+
+  @observable
+  public latestBlock: number = 0;
+
+  @observable
+  public heartbeats: SlotHeartbeatsData[];
+
+  constructor(keepPeriodsLimit = 15) {
+    this.keepPeriodsLimit = keepPeriodsLimit;
+
+    reaction(
+      () => operatorStore.slots,
+      () => {
+        this.storageKey = `hb-${operatorStore.address}`;
+        this.heartbeats = this.initHeartbeatsData();
+        operatorStore.contract.methods
+          .heartbeatColor()
+          .call()
+          .then(color => {
+            this.heartbeatColor = color;
+            this.loadHeartbeats();
+          });
+      }
+    );
+
+    autorun(this.storeData);
+  }
+
+  private initHeartbeatsData() {
+    let storedData = storage.load(this.storageKey);
+    storedData = storedData.length ? storedData : [];
+
+    return [
+      // add stored slots if any
+      ...storedData.map(s => ({ ...s, periods: new Set(s.periods) })),
+      // add new slots without heartbeat data if needed
+      ...range(storedData.length, operatorStore.slots.length - 1).map(i => ({
+        periods: new Set(),
+        lastPeriodChecked: 0,
+        owner: operatorStore.slots[i].owner,
+      })),
+    ];
+  }
+
+  @autobind
+  private storeData() {
+    if (!this.heartbeats) {
+      return;
+    }
+    storage.store(this.storageKey, toJS(this.heartbeats));
+  }
+
+  private async loadHeartbeatPeriodsRecursive(
+    slotHeartbeatsData: SlotHeartbeatsData,
+    heartbeatTxHash: Buffer
+  ): Promise<Set<number>> {
+    const txData = (await web3PlasmaStore.instance.eth.getTransaction(
+      bufferToHex(heartbeatTxHash)
+    )) as LeapTransaction;
+
+    const notHeartbeatTx = txData.from !== txData.to;
+    if (notHeartbeatTx) {
+      return;
+    }
+
+    const periodForTx = Math.floor(txData.blockNumber / 32) + 1;
+
+    const alreadySeenTheHeartbeat =
+      slotHeartbeatsData.periods.has(periodForTx) &&
+      slotHeartbeatsData.periods.has(periodForTx - 1);
+
+    if (alreadySeenTheHeartbeat) {
+      return;
+    }
+
+    slotHeartbeatsData.periods.add(periodForTx);
+
+    const heartbeatTrackingLimitReached =
+      this.keepPeriodsLimit &&
+      slotHeartbeatsData.periods.size > this.keepPeriodsLimit;
+
+    if (heartbeatTrackingLimitReached) {
+      return;
+    }
+
+    // recurse to the previous spend of heartbeat token, if any
+    const tx = Tx.fromRaw(txData.raw);
+    if (!tx.inputs.length) {
+      return;
+    }
+    return this.loadHeartbeatPeriodsRecursive(
+      slotHeartbeatsData,
+      tx.inputs[0].prevout.hash
+    );
+  }
+
+  private async loadHeartbeatsForSlot(slotId) {
+    const [heartbeatUtxo] = await web3PlasmaStore.instance.getUnspent(
```

Comment:
I thought we've disabled this for heartbeat color... 🤔

Reply:
that's a great catch! Totally overlooked — leap-sandbox runs a node with filtering switched off.

Reply:
I will fix this. Thanks for the catch

Reply:
addressed in latest commit. The only idea I can come up is to scan blocks a-la dapparatus/burner-wallet. Btw, it is so easy to bypass the filtering and push your heartbeat even without running the node at all.

Reply:
Yap, that what I discovered as well - sorry if I forgot to mention it 🙇‍♂️
I was thinking to wait for new period, then some more blocks after it and then parse them and find those that look like heartbeat. Assuming that validator will make heartbeat tx in first couple of blocks 🤷‍♂️